src/python_routing_v02/troute/routing/fast_reach/mc_reach.pyx

```diff
@@ -1200,8 +1247,8 @@ cpdef object compute_network_structured(
     cdef float reservoir_outflow, reservoir_water_elevation
     cdef int id = 0
     #Run time
-    with nogil:
-      while timestep < nsteps+1:
+    #with nogil:
```

Comment:
We have to support the nogil statement.

Reply:
#alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                            starting_bias_array[gage_i] = usgs_values[gage_i, timestep-1]
                                                                    ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1328:69: Converting to Python object not allowed without gil


I am getting a lot of these errors when trying to run with nogil. Is there a simple workaround for cython? The assigning has to happen.

Reply:
A large number of errors occurring once Gil is shut off. 

Error compiling Cython file:
------------------------------------------------------------
...
                        
            if gages_size:

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                   ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1318:20: Assignment of Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1327:43: Assignment of Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                            starting_bias_array[gage_i] = usgs_values[gage_i, timestep-1]
                                              ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1328:47: Assignment of Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:35: Truth-testing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:35: Calling gil-requiring function not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:29: Accessing Python attribute not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:27: Accessing Python global or builtin not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:35: Constructing Python tuple not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...

                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                                              ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1320:47: Converting to Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                                                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1321:67: Coercion from Python not allowed without the GIL

Error compiling Cython file:
------------------------------------------------------------
...
                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                                                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1321:67: Operation not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1321:59: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                for gage_i in range(gages_size):                        
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                                                                   ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1321:68: Operation not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                                                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:163: Coercion from Python not allowed without the GIL

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                                                                  ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:163: Operation not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                                                      ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:151: Operation not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                      ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:103: Operation not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                             ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:94: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                    ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:117: Converting to Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                                                        ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:153: Converting to Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                    usgs_position_i = usgs_positions_list[gage_i]
                    a = 120  # TODO: pull this a value from the config file somehow  
                    for i in range(0,len(usgs_values[0])):
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                                                                                                                                                                                       ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1322:184: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1325:43: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1325:43: Calling gil-requiring function not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                        if np.isnan(usgs_values[gage_i, timestep-1]):
                            da_weight = exp(decay_timesteps[gage_i]/-a)
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1325:43: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                            flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * da_weight + starting_bias_array[gage_i]  
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                                          ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1327:43: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                            starting_bias_array[gage_i] = usgs_values[gage_i, timestep-1]
                                              ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1328:47: Indexing Python object not allowed without gil

Error compiling Cython file:
------------------------------------------------------------
...
                            #alternative version below using decay_timestep not da_weight calculation based on discussions 
                            #flowveldepth[usgs_position_i, timestep , 0] = (starting_bias_array[gage_i] - flowveldepth[usgs_position_i, timestep , 0])  * decay_timesteps[gage_i] + starting_bias_array[gage_i]  
                            decay_timesteps[gage_i] += 1
                        else:
                            decay_timesteps[gage_i] = 1
                            starting_bias_array[gage_i] = usgs_values[gage_i, timestep-1]
                                                                    ^
------------------------------------------------------------

troute/routing/fast_reach/mc_reach.pyx:1328:69: Converting to Python object not allowed wi